src/get-static-value.mjs

```diff
@@ -266,13 +273,27 @@ function isEffectivelyConst(variable) {
     return false
 }
 
+/**
+ * @callback VisitorCallback
+ * @param {import('./types.mjs').Node} node
+ * @param {import('eslint').Scope.Scope|undefined} initialScope
+ * @returns {StaticValue | null}
+ */
+
+/** @type {Readonly<Partial<Record<import('eslint').Rule.NodeTypes, VisitorCallback>>> } */
 const operations = Object.freeze({
     ArrayExpression(node, initialScope) {
+        if (node.type !== "ArrayExpression") {
+            return null
+        }
```

Comment:
🤔 are these `if`s actually necessary, or just to appease a type checker that doesn't understand what's happening?

If it's the latter, the general right thing to do would be to fix TypeScript's understanding of the types so it doesn't yell at you.

For posterity: I tried adding `/** @param {import("estree").ArrayExpression} node */` to the `ArrayExpression` one and got a type error on `operations`:

```plaintext
Type 'Readonly<{ ArrayExpression(node: ArrayExpression, initialScope: Scope | undefined): { value: any[]; } | null; AssignmentExpression(node: Node, initialScope: Scope | undefined): StaticValue | null; ... 14 more ...; UnaryExpression(node: Node, initialScope: Scope | undefined): { ...; } | ... 3 more ... | null; }>' is not assignable to type 'Readonly<Partial<Record<"CatchClause" | "ClassBody" | "Identifier" | "Literal" | "MethodDefinition" | "PrivateIdentifier" | "Program" | "Property" | "PropertyDefinition" | "SpreadElement" | ... 60 more ... | "VariableDeclaration", VisitorCallback>>>'.
  Types of property 'ArrayExpression' are incompatible.
    Type '(node: ArrayExpression, initialScope: Scope | undefined) => { value: any[]; } | null' is not assignable to type 'VisitorCallback'.ts(2322)
```

Tricky.

I got around it by:
1. Renaming `types.mjs` to `types.d.ts`, so I could use "real" TypeScript syntax and not the annoying JSDoc sludge
2. Making the `VisitorCallback` type generic:
  ```ts
  export type VisitorCallback<InNode extends Node> = (
      node: InNode,
      initialScope: eslint.Scope.Scope | undefined,
  ) => StaticValue | null
  ```
3. Using a type annotation on the `ArrayExpression` member:
  ```ts
  /** @type {Readonly<Partial<Record<import('eslint').Rule.NodeTypes, import("./types.js").VisitorCallback<any>>>>} */
  const operations = Object.freeze({
      /** @type {import("./types.js").VisitorCallback<import("estree").ArrayExpression>} */
      ArrayExpression(node, initialScope) {
  ```

YMMV.